File: website/src/games/trikiUIMath.ts
/** Port `UI/TrikiUI/TrikiUIMath.swift` */
const NEUTRAL_ENTER = 0.12;
const NEUTRAL_EXIT = 0.08;

export function slotIndex(posX: number, slots: number): number {
  if (slots <= 1) return 0;
  const clamped = Math.min(1, Math.max(-1, posX));
  const slot = Math.floor(((clamped + 1) / 2) * slots);
  return Math.min(slots - 1, Math.max(0, slot));
}

export function focusedSlot(
  posX: number,
  slots: number,
  currentFocus: number | null = null,
): number | null {
  if (slots <= 0) return null;
  const neutralBand = currentFocus === null ? NEUTRAL_ENTER : NEUTRAL_EXIT;
  if (slots === 1) {
    return Math.abs(posX) < neutralBand ? null : 0;
  }
  if (Math.abs(posX) < neutralBand) return null;
  return slotIndex(posX, slots);
}

File: app/UI/TrikiUI/TrikiUIMath.swift
import Foundation

/// Maps Triki pointer movement to focusable UI slots.
///
/// Purpose: centralize slot-index math with neutral-zone hysteresis.
/// Use when: converting `posX` motion samples into stable menu focus.
/// Example: `TrikiUIMath.focusedSlot(posX: input.posX, slots: 4)`.
enum TrikiUIMath {
  /// Indeks slotu 0…(slots−1) z pozycji poziomej (−1…1).
  static func slotIndex(posX: Double, slots: Int) -> Int {
    guard slots > 1 else { return 0 }
    let clamped = min(1, max(-1, posX))
    let slot = Int(floor((clamped + 1) / 2 * Double(slots)))
    return min(slots - 1, max(0, slot))
  }

  /// Slot albo `nil` w strefie neutralnej (środek).
  /// Histereza: łatwiej utrzymać fokus niż go zdobyć (mniej mrugania przy drganiach).
  static func focusedSlot(
    posX: Double,
    slots: Int,
    currentFocus: Int? = nil,
    neutralEnterBand: Double = TrikiUIConfig.neutralEnterBand,
    neutralExitBand: Double = TrikiUIConfig.neutralExitBand
  ) -> Int? {
    guard slots > 0 else { return nil }
    let neutralBand = currentFocus == nil ? neutralEnterBand : neutralExitBand
    if slots == 1 {
      return abs(posX) < neutralBand ? nil : 0
    }
    if abs(posX) < neutralBand { return nil }
    return slotIndex(posX: posX, slots: slots)
  }
}
